File: youtube/models/watermarks.py
"""Pydantic models for the YouTube Watermarks resource.

Reference: https://developers.google.com/youtube/v3/docs/watermarks
"""

from __future__ import annotations

from enum import StrEnum

from youtube.models.base import YouTubeBaseModel

# ---------------------------------------------------------------------------
# Enumerations
# ---------------------------------------------------------------------------

class WatermarkTimingType(StrEnum):
    """When the watermark appears relative to the video."""

    OFFSET_FROM_START = "offsetFromStart"
    OFFSET_FROM_END = "offsetFromEnd"


class WatermarkCornerPosition(StrEnum):
    """Corner position for the watermark overlay."""

    TOP_LEFT = "topLeft"
    TOP_RIGHT = "topRight"
    BOTTOM_LEFT = "bottomLeft"
    BOTTOM_RIGHT = "bottomRight"


# ---------------------------------------------------------------------------
# Resource parts
# ---------------------------------------------------------------------------

class WatermarkTiming(YouTubeBaseModel):
    """Timing configuration for a watermark."""

    type: WatermarkTimingType | None = None
    offset_ms: str | None = None
    duration_ms: str | None = None


class WatermarkPosition(YouTubeBaseModel):
    """Position configuration for a watermark."""

    type: str | None = None
    corner_position: WatermarkCornerPosition | None = None


class Watermark(YouTubeBaseModel):
    """A YouTube Watermark resource (request body for ``watermarks.set``)."""

    timing: WatermarkTiming | None = None
    position: WatermarkPosition | None = None
    image_url: str | None = None
    image_bytes: str | None = None
    target_channel_id: str | None = None

File: go/watermarks.go
package youtube

import (
	"context"
	"errors"
	"net/url"
)

// Watermark timing type constants.
const (
	WatermarkTimingOffsetFromStart = "offsetFromStart"
	WatermarkTimingOffsetFromEnd   = "offsetFromEnd"
)

// Watermark corner position constants.
const (
	WatermarkCornerTopLeft     = "topLeft"
	WatermarkCornerTopRight    = "topRight"
	WatermarkCornerBottomLeft  = "bottomLeft"
	WatermarkCornerBottomRight = "bottomRight"
)

// WatermarkTiming is the timing configuration for a watermark.
type WatermarkTiming struct {
	Type       string `json:"type,omitempty"`
	OffsetMs   string `json:"offsetMs,omitempty"`
	DurationMs string `json:"durationMs,omitempty"`
}

// WatermarkPosition is the position configuration for a watermark.
type WatermarkPosition struct {
	Type           string `json:"type,omitempty"`
	CornerPosition string `json:"cornerPosition,omitempty"`
}

// Watermark is a YouTube Watermark resource (configuration for SetWatermark).
type Watermark struct {
	Timing          *WatermarkTiming   `json:"timing,omitempty"`
	Position        *WatermarkPosition `json:"position,omitempty"`
	ImageURL        string             `json:"imageUrl,omitempty"`
	ImageBytes      string             `json:"imageBytes,omitempty"`
	TargetChannelID string             `json:"targetChannelId,omitempty"`
}

// SetWatermarkParams configures SetWatermark.
type SetWatermarkParams struct {
	// ChannelID is the channel to set the watermark on. Required.
	ChannelID string
	// Watermark is the watermark configuration (timing, position). Mirrors the
	// Python SDK, which transmits the image via the binary media upload; the
	// configuration is accepted for parity.
	Watermark Watermark
	// ImageData is the raw image bytes (JPEG or PNG, max 10 MB). Required.
	ImageData []byte
	// ContentType is the image MIME type. Defaults to "image/png".
	ContentType string
}

// SetWatermark uploads and sets a watermark for a channel via a binary media
// upload. Mirrors watermarks.set. Requires OAuth.
func (c *Client) SetWatermark(ctx context.Context, p SetWatermarkParams) error {
	if err := c.requireOAuth(); err != nil {
		return err
	}
	if p.ChannelID == "" {
		return errors.New("youtube: SetWatermark requires a ChannelID")
	}
	if len(p.ImageData) == 0 {
		return errors.New("youtube: SetWatermark requires ImageData")
	}
	contentType := p.ContentType
	if contentType == "" {
		contentType = "image/png"
	}
	q := url.Values{}
	q.Set("channelId", p.ChannelID)
	q.Set("uploadType", "media")

	return c.uploadMedia(ctx, "watermarks/set", q, p.ImageData, contentType, nil)
}

// UnsetWatermark removes the watermark from a channel. Mirrors
// watermarks.unset. Requires OAuth.
func (c *Client) UnsetWatermark(ctx context.Context, channelID string) error {
	if channelID == "" {
		return errors.New("youtube: UnsetWatermark requires a ChannelID")
	}
	q := url.Values{}
	q.Set("channelId", channelID)
	return c.doPostNoContent(ctx, "watermarks/unset", q)
}
